### PR TITLE
Sky model and spec flexure for bkg_redux 

### PR DESCRIPTION
**Pull request by debora-pe**
This PR deals with cases of dithered observations (i.e., A-B reductions). Currently, in this scenario, the sky spectrum generated by PypeIt it's not actually a sky spectrum (because the sky is removed when we subtract the B frame) so there is no 2D/1D sky spectrum available and if the spectral flexure correction is required it does not work.
Here, the sky model is performed using the A frames and saved. Specifically:
- The spec2d now includes an additional image of the actual sky model, called `nobkg_skymodel` (`skymodel` still exists)
- The 1d sky spectrum is now extracted from `nobkg_skymodel`
- Spectral flexure is using the 1d sky spectrum extracted from `nobkg_skymodel`

Note: `nobkg_skymodel` does not exist if the observations are not dithered.

### CODE REVIEW

File: pypeit/core/flexure.py [hunk 1]
```diff
@@ -238,25 +239,43 @@ def spec_flex_shift(obj_skyspec, arx_skyspec, arx_fwhm_pix, spec_fwhm_pix=None,
...
-    obj_sky_flux = obj_skyspec.flux.value - obj_sky_cont
-    pypeitFit_sky, _ = fitting.iterfit(arx_skyspec.wavelength.value, arx_skyspec.flux.value,
-                                       nord = 3,  kwargs_bspline={'everyn': everyn}, kwargs_reject={'groupbadpix':True,'maxrej':1},
-                                       maxiter = 15, upper = 3.0, lower = 3.0)
-    arx_sky_cont, _ = pypeitFit_sky.value(arx_skyspec.wavelength.value)
-    arx_sky_flux = arx_skyspec.flux.value - arx_sky_cont
-
-    # Consider sharpness filtering (e.g. LowRedux)
-    msgs.work("Consider taking median first [5 pixel]")
+    # msgs.work("Consider taking median first [5 pixel]")
```
Comment (profxj) [inline pypeit/core/flexure.py]: remove?

File: pypeit/core/flexure.py [hunk 2]
```diff
@@ -238,25 +239,43 @@ def spec_flex_shift(obj_skyspec, arx_skyspec, arx_fwhm_pix, spec_fwhm_pix=None,
...
+    # obj_sky_cont, _ = pypeitFit_obj.value(obj_skyspec.wavelength.value)
+    #
+    # obj_sky_flux = obj_skyspec.flux.value - obj_sky_cont
+    # pypeitFit_sky, _ = fitting.iterfit(arx_skyspec.wavelength.value, arx_skyspec.flux.value,
+    #                                    nord = 3,  kwargs_bspline={'everyn': everyn}, kwargs_reject={'groupbadpix':True,'maxrej':1},
+    #                                    maxiter = 15, upper = 3.0, lower = 3.0)
+    # arx_sky_cont, _ = pypeitFit_sky.value(arx_skyspec.wavelength.value)
+    # arx_sky_flux = arx_skyspec.flux.value - arx_sky_cont
+
+    # DP: ANY OBJECTION to changing the above to the following? It seems working much better.
```
Comment (profxj) [inline pypeit/core/flexure.py]: Very sensible!

File: pypeit/find_objects.py [hunk 3]
```diff
@@ -505,6 +505,11 @@ def global_skysub(self, skymask=None, update_crmask=True,
         Args:
             skymask (`numpy.ndarray`_, optional):
                 A 2D image indicating sky regions (1=sky)
+            nobkg_sciimg (:obj:`dict`, optional):
```
Comment (profxj) [inline pypeit/find_objects.py]: not the cleanest to use a `dict`, but at least it is a simple object :)

Comment (profxj) [reply pypeit/find_objects.py]: Can you please specify the required keys?

File: pypeit/pypeit.py [hunk 4]
```diff
@@ -734,6 +736,11 @@ def objfind_one(self, frames, det, bg_frames=None, std_outfile=None):
             List of objects found
         sciImg : :class:`~pypeit.images.pypeitimage.PypeItImage`
             Science image
+        nobkg_sciimg : :obj:`dict`
+            Dictionary containing the science image and ivar
```
Comment (profxj) [inline pypeit/pypeit.py]: same as above -- list the required `keys`

File: pypeit/pypeit.py [hunk 5]
```diff
@@ -920,6 +934,11 @@ def extract_one(self, frames, det, sciImg, objFind, initial_sky, sobjs_obj):
             sciImg (:class:`~pypeit.images.pypeitimage.PypeItImage`):
                 Data container that holds a single image from a
                 single detector its related images (e.g. ivar, mask)
+            nobkg_sciimg (:obj:`dict`, optional):
```
Comment (profxj) [inline pypeit/pypeit.py]: same

**Review by profxj — APPROVED** [review]
A bit klunky to squeeze this in, but I'm not seeing another option..

Maybe Kyle will have an idea or two

Am approving, but please do address my minor comments

**Comment by profxj** [comment]
If NIRSPEC is going to bump the version (new datamodel), we might 
push this and its companion through too.

File: pypeit/core/flexure.py (hunk 2 above)
Comment (kbwestfall) [reply pypeit/core/flexure.py]: FWIW, looks good to me, too.

File: pypeit/find_objects.py (hunk 3 above)
Comment (kbwestfall) [reply pypeit/find_objects.py]: Agreed.  Can we make it a `PypeItImage`?

**Review by kbwestfall — COMMENTED** [review]
I have two likely annoying comments that I wanted to bring up before approving:

1. The structural comment is that I think it would be cleaner if we didn't alter the core functions.  What I mean is that, e.g., the only reason `nobkg_skyimg` is added to the calling function of `extract_optimal` is so that, if it exists, it will use it instead of `skyimg` (if I'm reading the code correctly).  I think we should handle this in the calling function instead.  I.e., instead of `extract_optimal(..., skyimg, nobkg_skyimg=nobkg_skyimg, ...)`, can we instead do `extract_optimal(..., skyimg if nobkg_skyimg is None else nobkg_skyimg, ...)`?  The latter doesn't require any changes to the `extract_optimal` function.

2. I find the nomenclature choice confusing, but that might only be me.  To me, `nobkg_skyimg` sounds like a sky image that doesn't have any background.  I don't have a great alternative suggestion, though.  The only thing that came to mind is the following.  When difference imaging, the `skyimg` is actually a residual of the two images, and what you're calling `nobkg_skyimg` is really the sky image.  Is it too much of a hassle to change up the naming system to match this?  I.e., when difference imaging, change `skyimg` to something like `skyresid` and change `nobkg_skyimg` to `skyimg`?

**Comment by debora-pe** [comment]
> I have two likely annoying comments that I wanted to bring up before approving:
> 
> 1. The structural comment is that I think it would be cleaner if we didn't alter the core functions.  What I mean is that, e.g., the only reason `nobkg_skyimg` is added to the calling function of `extract_optimal` is so that, if it exists, it will use it instead of `skyimg` (if I'm reading the code correctly).  I think we should handle this in the calling function instead.  I.e., instead of `extract_optimal(..., skyimg, nobkg_skyimg=nobkg_skyimg, ...)`, can we instead do `extract_optimal(..., skyimg if nobkg_skyimg is None else nobkg_skyimg, ...)`?  The latter doesn't require any changes to the `extract_optimal` function.
> 2. I find the nomenclature choice confusing, but that might only be me.  To me, `nobkg_skyimg` sounds like a sky image that doesn't have any background.  I don't have a great alternative suggestion, though.  The only thing that came to mind is the following.  When difference imaging, the `skyimg` is actually a residual of the two images, and what you're calling `nobkg_skyimg` is really the sky image.  Is it too much of a hassle to change up the naming system to match this?  I.e., when difference imaging, change `skyimg` to something like `skyresid` and change `nobkg_skyimg` to `skyimg`?



Hi @kbwestfall , thank you so much for your feedback. I have a couple of comments back.
For 1., that was actually my original approach, but I then saw that these lines of code
```
    # TODO This makes no sense for difference imaging? Not sure we need NIVAR anyway
    var_no = None if base_var is None \
                else procimg.variance_model(base_var, counts=skyimg, count_scale=count_scale,
                                            noise_floor=noise_floor)
```

are present both in `extract_optimal` and `extract_boxcar`, so I did not wanted to mess up the variance model for the extracted object. It is true that the inline comment above the code says that this does not make sense for difference imaging, but I was not confident enough to make some change here. Do you have any suggestion?

For 2., I agree that the naming is a bit confusing, but I did not want to change much of the existing nomenclature. Would `bkg_skyimg` or `bkg_redux_skyimg` work? The second one looks a bit too long.

About your comment on using  `PypeItImage` instead of a `dict`, the reason I used `dict` is that I only need `image` and `ivar` and I was afraid that another `PypeItImage` in the reduction flow would slow down the reduction, especially for instruments with many detectors. What would you suggest? Maybe I could create a `PypeItImage` that has only `image` and `ivar`, would that work?

**Comment by kbwestfall** [comment]
> Hi @kbwestfall , thank you so much for your feedback. I have a couple of comments back. For 1., that was actually my original approach, but I then saw that these lines of code
> 
> ```
>     # TODO This makes no sense for difference imaging? Not sure we need NIVAR anyway
>     var_no = None if base_var is None \
>                 else procimg.variance_model(base_var, counts=skyimg, count_scale=count_scale,
>                                             noise_floor=noise_floor)
> ```
> 
> are present both in `extract_optimal` and `extract_boxcar`, so I did not wanted to mess up the variance model for the extracted object. It is true that the inline comment above the code says that this does not make sense for difference imaging, but I was not confident enough to make some change here. Do you have any suggestion?

Good catch, but I agree with the TODO; i.e, when performing difference imaging, we should be passing `nobkg_skyimg` to `variance_model`, not `skyimg`.  So I think your original approach is better.  This will definitely have an effect on the noise modeling when difference imaging, but I think it is the correct change to make.  Although, because it's difference imaging, there may be a sqrt(2) (assuming A-B) in the noise that we should make sure is being accounted for (but maybe this is rehashing something we/you already worked through with MOSFIRE...). 

> For 2., I agree that the naming is a bit confusing, but I did not want to change much of the existing nomenclature. Would `bkg_skyimg` or `bkg_redux_skyimg` work? The second one looks a bit too long.

Of the two, I actually like the 2nd one better, but I think both are an improvement.

> About your comment on using `PypeItImage` instead of a `dict`, the reason I used `dict` is that I only need `image` and `ivar` and I was afraid that another `PypeItImage` in the reduction flow would slow down the reduction, especially for instruments with many detectors. What would you suggest? Maybe I could create a `PypeItImage` that has only `image` and `ivar`, would that work?

I took a quick look back at `PypeItImage` and, while it should be possible to instantiate a `PypeItImage` with only `image` and `ivar`, it does feel like overkill.  All the other elements of the datamodel should be None, except for `fullmask`.  But even allocating memory for `fullmask` could lead to a performance hit for large images.

I had another thought:  In `pypeit.py` where `nobkg_sciimg` is defined, `bgimg` is also defined.  It's also worth noting that `sciImg.sub` creates (allocates) a new `PypeItImage`.  If feels like there's some combination of either passing around `bgimg` or redefining the name of `sciImg` so that we avoid any additional memory allocation.  Would something like this work?
```
if bg_frames ... :
    nobkg_sciimg = sciImg
    bg_file_list = ...
    bgimg = ...
    sciImg = nobkg_sciimg.sub(bgimg)
```

**Comment by debora-pe** [comment]
Thank you @kbwestfall for all your suggestions. I have implemented them. I'm now running the test and I'll see how it goes.

**Comment by debora-pe** [comment]
Tests pass.
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  247 passed, 65 warnings in 368.44s (0:06:08) ---
--- PYTEST UNIT TESTS PASSED  133 passed, 170 warnings in 1833.67s (0:30:33) ---
--- PYTEST VET TESTS PASSED  61 passed, 96 warnings in 3937.48s (1:05:37) ---
--- PYPEIT DEVELOPMENT SUITE PASSED 237/237 TESTS  ---
Testing Started at 2024-02-26T21:03:53.271145
Testing Completed at 2024-02-27T10:04:39.559941
Total Time: 13:00:46.288796
```

File: pypeit/core/flexure.py [hunk 6]
```diff
@@ -217,56 +211,59 @@ def spec_flex_shift(obj_skyspec, arx_skyspec, arx_fwhm_pix, spec_fwhm_pix=None,
     obj_skyspec.data['flux'][0,:] = np.maximum(obj_skyspec.data['flux'][0,:], 0.)
     arx_skyspec.data['flux'][0,:] = np.maximum(arx_skyspec.data['flux'][0,:], 0.)
 
+    # clip too large values (>90%) only in obj_skyspec (assuming arx_skyspec is being vetted before)
+    # this is used ony for the cross-correlation
+    obj_skyspec_flux = obj_skyspec.flux.value
+    _lower = np.percentile(obj_skyspec_flux[obj_skyspec_flux < 0.0], 90) if np.any(obj_skyspec_flux < 0.0) else 0.0
+    _upper = np.percentile(obj_skyspec_flux[obj_skyspec_flux >= 0.0], 90) if np.any(obj_skyspec_flux >= 0.0) else 0.0
+    obj_skyspec_flux = np.clip(obj_skyspec_flux, _lower, _upper)
```
Comment (kbwestfall) [inline pypeit/core/flexure.py:225]: You perform a similar operation a few times, it's likely worth making a short function.

File: pypeit/core/extract.py [hunk 7]
```diff
@@ -377,7 +379,7 @@ def extract_boxcar(sciimg, ivar, mask, waveimg, skyimg, spec, fwhmimg=None, base
         added.
     """
     # Setup
-    imgminsky = sciimg - skyimg
+    # imgminsky = sciimg - skyimg
```
Comment (kbwestfall) [inline pypeit/core/extract.py:382]: Nice!

**Review by kbwestfall — APPROVED** [review]
Looks good!  One very minor comment.

**Comment by debora-pe** [comment]
Merged in #1775 and #1772 
Re-running the tests now.

**Comment by debora-pe** [comment]
Still all good.
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  249 passed, 65 warnings in 356.83s (0:05:56) ---
--- PYTEST UNIT TESTS PASSED  133 passed, 170 warnings in 1686.85s (0:28:06) ---
--- PYTEST VET TESTS PASSED  61 passed, 97 warnings in 5188.37s (1:26:28) ---
--- PYPEIT DEVELOPMENT SUITE PASSED 237/237 TESTS  ---
Testing Started at 2024-02-27T19:45:16.135822
Testing Completed at 2024-02-28T08:49:26.571723
Total Time: 13:04:10.435901
```

Doc updated
```
The HTML pages are in _build/html.

Build finished. The HTML pages are in _build/html.


========================
Sphinx build successful.
========================
```

**Comment by debora-pe** [comment]
Merging

**Pull request merged**